runtime/gc_structs/ClassArrayClassSlotIterator.cpp

```diff
@@ -46,10 +46,27 @@ GC_ClassArrayClassSlotIterator::nextSlot()
 			classPtr = (J9Class *)_iterateClazz->arrayClass;
 			if (!_isArrayClass) {
 				_state = classArrayClassSlotIterator_state_done;
+				break;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if (NULL == classPtr) {
+				classPtr = (J9Class *)_iterateClazz->nullRestrictedArrayClass;
+				_state += classArrayClassSlotIterator_state_componentType;
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
 			} else {
 				_state += 1;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+				if (NULL == ((J9ArrayClass *)classPtr)->companionArray) {
+					_state += 1; /* skip nullRestrictedArrayClass case */
+				}
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```

Comment:
This block is in a wrong place. This check should be done inside `classArrayClassSlotIterator_state_nullRestrictedArrayClass` state.